### ISSUE TITLE
Switch to isomorphic-fetch

### ISSUE THREAD

**Issue by nicholaswyoung** [heroku/react-refetch#11]
The module, as it sits, is almost perfect for my use case. However, I need to render components server side too. Would you be open to a PR that replaces `whatwg-fetch` with a similar, standards compliant polyfill that also works in Node?

**Comment by ryanbrainard** [comment]
I'm not opposed to it, but just wondering if it would be better to allow applications to provide their own `fetch` polyfill. Perhaps it should just be a dev dependency in this library and then it would be up to applications to choose what works best for them? Also, `whatwg-fetch` will not load if you already have another `fetch` loaded already, so you could probably still use it as-is today and just make sure your `fetch` is loaded first. I do not claim to be an expert at all on JavaScript modules, imports, and polyfills, so  I'm all ears if someone has a better idea how best to do this.

**Comment by nicholaswyoung** [comment]
@ryanbrainard That's an even better idea. The big issue of mine was that Node doesn't recognize `window`, so we'll need to decide how that dependency is passed in. That's what originally led me down this path.

I'll dig into this and see if I can come up with a workable solution.

**Comment by wilmoore** [comment]
@nicholaswyoung 

> ...The big issue of mine was that Node doesn't recognize window, so we'll need to decide how that dependency is passed in. That's what originally led me down this path.

You might check out: https://www.npmjs.com/package/global

**Comment by hnordt** [comment]
@ryanbrainard I'm just wondering why `whatwg-fetch` is a peerDependency and not a dependency?

**Comment by ryanbrainard** [comment]
@hnordt The idea was to allow applications to have more freedom to choose which version of `whatwg-fetch` so it would not conflict if it was already a dependency of the application. Although, to be honest, I'm kind of an npm noob, so if someone knows a better way to do this, please let me know. If we can get away with just making it a devDependency for tests and have applications use whatever polyfill they want (or nothing at all if they don't care about cross-browser support) in production would be ideal. How do other npm modules usually deal with polyfill dependencies like this?

**Comment by wilmoore** [comment]
> How do other npm modules usually deal with polyfill dependencies like this?

Other npm packages omit the dependency from `dependencies`, but put it in `devDependencies` for tests. The user is then responsible for installing the actual dependency into their own project.

**Comment by ryanbrainard** [comment]
> Other npm packages omit the dependency from dependencies, but put it in devDependencies for tests. The user is then responsible for installing the actual dependency into their own project.

That sounds good to me. I think that would fix @nicholaswyoung 's original problem as well so he can use whatever  `fetch` he wants. One additional problem that might come up after this is the use of `window.fetch` and `window.Request`. What do people usually do for Node interoperability of these?

**Comment by hnordt** [comment]
@ryanbrainard another problem I see is that for some apps, the API URL is different from the app URL.

For example, app can be running on `http://foo.com` but the API is running on `http://api.foo.com`.

Specifying URL for every mapping isn't a good option I think.

For some apps we need to work with tokens too.

Maybe you can install `whatwg-fetch` as a dependency, but enable us to implement our own version of it.

For example:

```js
import SomeFetchImplementation from 'some-fetch-impleentation';

const MyOwnFetchImplementation = (method, url, params) => {
  return SomeFetchImplementation(method, url, params, {
    Authorization: localStorage.token ? `Token: ${localStorage.token}` : ''
  }).then(response => {
    if (url === '/auth/login') {
      localStorage.token = response.body.token;
    }
    return response.body;
  }).catch(error => {
    if (error.status === UNAUTHORIZED) {
      alert('Session expired');
      localStorage.token = null;
    }
  });
};

ReactRefetch.fetch = MyOwnFetchImplementation;
```

Or maybe if we just want to specify a global URL:

```js
RectRefetch.fetch.URL = 'http://api.foo.com';
```

**Comment by ryanbrainard** [comment]
@hnordt I like the idea of letting applications provide their own `fetch`; however, it seems like doing things like default URLs and custom token management do not belong in there. I agree we probably need some mechanism for that, but just doesn't seem like it belongs in a `fetch` implementation. Instead, I think this could be solved with a custom request preprocessor (see https://github.com/heroku/react-refetch/issues/9) that runs before `fetch`.

**Comment by hnordt** [comment]
@ryanbrainard agreed. +1 for custom request preprocessor for the other cases I specified

**Comment by hnordt** [comment]
@ryanbrainard I don't know if it's good yet, but if you let applications provide their own `fetch` it open doors for people implementing offline support, we could save requests in IndexedDB, and refetch only if we have an active connection.

**Comment by nicholaswyoung** [comment]
@ryanbrainard I have a branch that is working with an alternative fetch implementation (in Node, server-side). It needs a bit of polish, but should be ready either tonight or tomorrow.

**Comment by ryanbrainard** [comment]
@nicholaswyoung sweet! Looking forward to to it. Just a heads up that I might slow to respond or get anything merged or published during the holidays.

**Comment by nicholaswyoung** [comment]
@ryanbrainard It's all good! Thanks for the heads up, but it's what I expect. (And what should be happening. Enjoy!)

**Comment by boyswan** [comment]
any news on this?

**Comment by nicholaswyoung** [comment]
Still working on perfecting it, @boyswan. It'll be coming around soon though.

**Comment by ryanbrainard** [comment]
@nicholaswyoung take your time :) just wondering, will you be moving fetch into devDependencies as suggested in https://github.com/heroku/react-refetch/issues/11#issuecomment-166035146 ? I think that's the way to go, but wondering what your approach is going to be. Of course, I'm open to other suggestions as well.

**Comment by nicholaswyoung** [comment]
That's exactly the idea, @ryanbrainard. I'm hoping to have the PR out this weekend (now that festivities are over). If I can knock out two issues with one commit, then I should at least try.

**Comment by nicholaswyoung** [comment]
Update: So, I thought finishing the PR would be easier than it was. I have the prototype working, but I'm still feeling out a couple of the design decisions, and want to make sure my code fits in nicely alongside what has already been written. *tldr: I have a solution, but it needs cleanup.*

I'll probably be tied up with client work for most of the week. I'd like to finish the PR, but know that it'll be a while. If someone gets to it first, I completely understand.

**Comment by hnordt** [comment]
@nicholaswyoung please share your code so we can improve it together. Better than starting from scratch :D

**Comment by ryanbrainard** [comment]
+1, just open it now as a "WIP" (work in progress) PR for early feedback.

**Comment by nicholaswyoung** [comment]
Will do when I get a chance to push it up.

**Comment by ahlee2326** [comment]
+1, this project is great but why only for client-side :'(

**Comment by ryanbrainard** [comment]
@nicholaswyoung Just wanted to check in on this. Anything I can help with on your PR?

**Comment by nicholaswyoung** [comment]
@ryanbrainard Catching up... a lot of things, this issue included, have fell by the wayside as I prepare for the impending arrival of a new baby. ;)

I don't think I have the bandwidth to tackle a configurable `fetch` dep, but I can easily submit a PR that switches to `node-fetch` when used server side. Would that at least fix the current issue? What do you think?

**Comment by ryanbrainard** [comment]
@nicholaswyoung No worries, the baby is definitely more important than this. Just push up what you have, add a comment that it is a work in progress, and someone else can always build on it. 

> I can easily submit a PR that switches to node-fetch when used server side. Would that at least fix the current issue? What do you think?

That sounds great. It is especially good if it is a stepping stone to a more general solution. Again, just push up what you have. If it works as is and adds incremental value, we can pull it in and improve on it later. If it needs work, I'm sure someone will step in and help.

Good luck with the new arrival! :baby: :baby_bottle:

**Comment by nicholaswyoung** [comment]
@ryanbrainard She isn't here yet, but is coming fast! (And one day, she will probably be annoyed that Dad brought her up on GitHub, but...)

In all of the hubbub, I just realized that I lost the changes I had, and didn't check them in. But, I'll try to get a PR in tomorrow morning.

**Comment by hnordt** [comment]
@ryanbrainard my suggestion on https://github.com/heroku/react-refetch/issues/40#issuecomment-172942228 can handle this.

**Comment by ryanbrainard** [comment]
@hnordt Your suggestion could technically work around this, but this is something that should be supported out of the box without users having to create their own handlers.

**Comment by hnordt** [comment]
@ryanbrainard following my idea I thought a isomorphicRequestEnhancer could be created to handle this:

```js
// enhancedConnect.js

import { connect }  from 'react-fetch';
import isomorphicRequestEnhancer from 'react-fetch-isomorphic-request-enhancer';
import myOtherRequestEnhancer from '<...>';

const enhancedConnect = compose(isomorphicRequestEnhancer, myOtherRequestEnhancer, connect);

export default enhancedConnect;
```

**Comment by nicholaswyoung** [comment]
I'm going to bow out, since I haven't yet had a chance to get down into the code this week. I'm covered up with too many other projects.

If @hnordt can solve this and assemble a PR, I'll back anyone's idea... Because it's better than mine at this stage!

**Comment by hnordt** [comment]
I'll not implement anything yet. We are on the discussing stage.

**Comment by ryanbrainard** [comment]
No problem. Thanks @nicholaswyoung and good luck with the :baby:

**Comment by ryanbrainard** [comment]
May be addressed in #75

**Comment by ryanbrainard** [comment]
Closed by #76 and release in [v1.0.0-beta.0](https://github.com/heroku/react-refetch/releases/tag/v1.0.0-beta.0)
